File: src/test/java/com/letgo/book/unit/application/find/FindBookQueryHandlerTest.java
package com.letgo.book.unit.application.find;

import com.letgo.book.application.find.FindBookQuery;
import com.letgo.book.application.find.FindBookQueryHandler;
import com.letgo.book.application.find.FindBookQueryResponse;
import com.letgo.book.domain.Book;
import com.letgo.book.domain.BookFinder;
import com.letgo.book.domain.BookNotFound;
import com.letgo.book.unit.application.BookTestCase;
import com.letgo.book.unit.domain.BookIdMother;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrowsExactly;

final public class FindBookQueryHandlerTest extends BookTestCase {
    private final FindBookQueryHandler handler = new FindBookQueryHandler(new BookFinder(repository));

    @Test
    public void itShouldFindABook() {
        Book book = anExistingBook();

        FindBookQuery query = new FindBookQuery(book.id().value());
        FindBookQueryResponse response = handler.handle(query);

        assertEquals(book.title().value(), response.getTitle());
    }

    @Test
    public void itShouldThrowAnExceptionWhenBookIsNotFound() {
        String id = BookIdMother.random().value();
        FindBookQuery query = new FindBookQuery(id);

        BookNotFound exception = assertThrowsExactly(
                BookNotFound.class,
                () -> handler.handle(query)
        );

        assertEquals("Book not found with id " + id, exception.getMessage());
    }
}

File: src/test/java/com/letgo/book/unit/application/find/FindBookQueryHandlerTest.kt
package com.letgo.book.unit.application.find

import com.letgo.book.application.BookResponse
import com.letgo.book.application.find.FindBookQuery
import com.letgo.book.application.find.FindBookQueryHandler
import com.letgo.book.domain.BookFinder
import com.letgo.book.domain.BookNotFound
import com.letgo.book.unit.application.BookTestCase
import com.letgo.book.unit.domain.ABookId
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrowsExactly
import org.junit.jupiter.api.Test

private class FindBookQueryHandlerTest : BookTestCase() {
    private val handler = FindBookQueryHandler(BookFinder(repository))

    @Test
    fun `It should find a book`() {
        val currentBook = anExistingBook()
        val id = currentBook.id().value()

        assertEquals(
            BookResponse(
                id = id,
                title = currentBook.title().value(),
                isEdited = currentBook.hasBeenEdited(),
            ),
            handler.handle(
                FindBookQuery(
                    id = id,
                )
            )
        )
    }

    @Test
    fun `It should throw an exception when book is not found`() {
        val id = ABookId.random().value()

        assertThrowsExactly(BookNotFound::class.java) {
            handler.handle(
                FindBookQuery(
                    id = id,
                )
            )
        }.run {
            assertEquals("Book not found with id $id", message)
        }
    }
}
